docs/mirgrations.md

```diff
@@ -0,0 +1,33 @@
+# State Migration Best Practices
+
+These best practices should apply to any team that contributes to an application that leverages local state.
+
+## Best Practices
+- When to Add a Migration: You should add a migration whenever you make a change to the shape of your state that is not backwards-compatible. This includes adding or removing properties, changing the type of a property, or moving properties around within the state tree (e.g. breaking controllers changes, etc...).
+- Detecting Errors in a Migration: The best way to detect errors in a migration is through thorough testing. Write tests that take various shapes of old state and ensure that they are correctly transformed into the new state shape. Also, use TypeScript or another type system to help catch type errors.
+- Handling Migration Errors: If a migration fails, you should have a strategy in place to handle the error. This could be as simple as logging the error and continuing with the default state, or it could involve more complex error recovery logic. In any case, it's important to ensure that your app can still function in some way even if a migration fails.
+
+- Example:
+```
+  if (
+    phishingControllerState?.hotlistLastFetched &&
+    phishingControllerState?.stalelistLastFetched
+  ) {
+    // This will make the list be fetched again when the user updates the app
+    state.engine.backgroundState.PhishingController.hotlistLastFetched = 0;
+    state.engine.backgroundState.PhishingController.stalelistLastFetched = 0;
+  } else {
+    captureException(
+      new Error(
+        `Migration 26: Invalid PhishingControllerState hotlist and stale list fetched: '${JSON.stringify(
+          state.engine.backgroundState.PhishingController,
+        )}'`,
+      ),
+    );
+  }
+
+  return state;
+```
+
+- Keeping Migrations Idempotent: Make sure your migrations are idempotent, meaning they can be run multiple times without changing the result beyond the initial application. This makes migrations safer and easier to test.
+- Removing Old Migrations: Over time, you may accumulate many migrations. At some point, it may be safe to remove old migrations, especially if you know that all users have migrated to a newer version of the state. Be cautious with this, as removing a migration could break the ability to upgrade for any users who are still on old versions of the state.
```

Comment:
This seems a bit contradictory to me. This suggests that it "may be safe" to remove old migrations, while acknowledging that it could break for some users.

There is no way for us to know whether any users have dormant wallet instances, so this is never a completely safe thing to do.

Reply:
Will we ever start to end of life application versions? This is done on the mobile side when we force upgrade. 

This may not apply now but could apply in the future. We can remove this section but it is something we should keep in mind.

Reply:
Hmm, so maybe we should clearly define these terms a bit more. I interpreted "safe" as meaning "won't break anything", which is different from this notion of which versions are "end-of-life" (i.e. no longer supported/considered).

For old version support, there are two separate cases to consider. First is which versions we support being actively used, and the second is which versions we support updating from. For active use, support is limited to just the versions that have working API integrations. I guess that's what your team is using the forced updates for, and that's likely what the extension team would do as well if they had this feature.

But at the moment we support upgrading from _any_ version, so that a user can update and things be in working order no matter the version that was last used. Dropping migration support for older versions would be a new thing, to my knowledge.

Reply:
We can certainly consider dropping migration support, but we don't really have a way to know how many affected wallets are out there. Any affected wallets would be dormant until the user wanted to update them, so we'd have no way to collect that data.

We should have a plan for how customer success handles these cases, if we are going to do something like that. We could have improved in-app support for that case as well, e.g. detect old versions and allow exporting the SRP and any keys (after the password is validated of course).

But it might be easier to leave them in-place for now, than it would be to make this a nice experience that wouldn't overly burden our CS team.